#Contents List
##Import Libraries
##Create Path
##Import Data - customers.csv
##Drop the date_joined column
##Rename columns for readability and intuitiveness
##Check for mixed types in df_customers
##Check for missing values
##Create a new dataframe to view missing values
##Addressing missing values
##rechecking missing values
##Look for duplicate records
##Import Data - orders_products_merged_updated_2.pkl
##Change the data type of user_id to string
##Change the data type of order_id to string
##Drop the _merge column from the df_ords_prods_merge dataframe as it is not needed. 
##Merge df_ords_prods_merge with df_customers
##Check the frequency of the new dataframe
##8. Export this new dataframe as a pickle file. - orders_products_all.pkl

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Path
path = r'C:\Users\Family Computer\Documents\Data Analytics\Instacart Basket Analysis'

In [3]:
#Import customers.csv
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
df_customers.shape

(206209, 10)

In [5]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
df_customers.head(100)

user_id  First Name    Surnam  Gender       STATE  Age date_joined  \
0     26711     Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1     33890    Patricia      Hart  Female  New Mexico   36    1/1/2017   
2     65803     Kenneth    Farley    Male       Idaho   35    1/1/2017   
3    125935    Michelle     Hicks  Female        Iowa   40    1/1/2017   
4    130797         Ann   Gilmore  Female    Maryland   26    1/1/2017   
..      ...         ...       ...     ...         ...  ...         ...   
95    86389     Stephen   Murillo    Male     Vermont   69    1/1/2017   
96   102956       Kathy  Mitchell  Female    Missouri   78    1/1/2017   
97   180919  Jacqueline     Roach  Female    Nebraska   21    1/1/2017   
98    74155       Karen      Berg  Female     Alabama   33    1/1/2017   
99    96166         NaN   Oconnor    Male    Oklahoma   51    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
..           ...                               ...     ...  
95             3                           married   66800  
96             1                           married   96314  
97             3  living with parents and siblings   56831  
98             3                           married   95285  
99             1                           married  155673  

[100 rows x 10 columns]

In [9]:
#Drop the date_joined column
df_customers = df_customers.drop(columns = ['date_joined'])

In [10]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age  n_dependants  \
0    26711    Deborah  Esquivel  Female    Missouri   48             3   
1    33890   Patricia      Hart  Female  New Mexico   36             0   
2    65803    Kenneth    Farley    Male       Idaho   35             2   
3   125935   Michelle     Hicks  Female        Iowa   40             0   
4   130797        Ann   Gilmore  Female    Maryland   26             1   

  fam_status  income  
0    married  165665  
1     single   59285  
2    married   99568  
3     single   42049  
4    married   40374

In [11]:
#Rename columns for readability and intuitiveness
df_customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [12]:
df_customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [13]:
df_customers.rename(columns = {'n_dependants' : 'Number_of_Dependants'}, inplace = True)

In [14]:
df_customers.rename(columns = {'fam_status' : 'Familial_Status'}, inplace = True)

In [15]:
df_customers.head(100)

user_id  First Name Last Name  Gender       State  Age  \
0     26711     Deborah  Esquivel  Female    Missouri   48   
1     33890    Patricia      Hart  Female  New Mexico   36   
2     65803     Kenneth    Farley    Male       Idaho   35   
3    125935    Michelle     Hicks  Female        Iowa   40   
4    130797         Ann   Gilmore  Female    Maryland   26   
..      ...         ...       ...     ...         ...  ...   
95    86389     Stephen   Murillo    Male     Vermont   69   
96   102956       Kathy  Mitchell  Female    Missouri   78   
97   180919  Jacqueline     Roach  Female    Nebraska   21   
98    74155       Karen      Berg  Female     Alabama   33   
99    96166         NaN   Oconnor    Male    Oklahoma   51   

    Number_of_Dependants                   Familial_Status  income  
0                      3                           married  165665  
1                      0                            single   59285  
2                      2                           married   99568  
3                      0                            single   42049  
4                      1                           married   40374  
..                   ...                               ...     ...  
95                     3                           married   66800  
96                     1                           married   96314  
97                     3  living with parents and siblings   56831  
98                     3                           married   95285  
99                     1                           married  155673  

[100 rows x 9 columns]

In [16]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Number_of_Dependants  206209 non-null  int64 
 7   Familial_Status       206209 non-null  object
 8   income                206209 non-null  int64 
dtypes: int64(4), object(5)
memory usage: 14.2+ MB


In [17]:
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [18]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  object
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Number_of_Dependants  206209 non-null  int64 
 7   Familial_Status       206209 non-null  object
 8   income                206209 non-null  int64 
dtypes: int64(3), object(6)
memory usage: 14.2+ MB


In [19]:
df_customers.describe()

Age  Number_of_Dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [20]:
# Check for mixed types in df_customers
for column in df_customers.columns:
    print(column,':',pd.api.types.infer_dtype(df_customers[column]))

user_id : string
First Name : string
Last Name : string
Gender : string
State : string
Age : integer
Number_of_Dependants : integer
Familial_Status : string
income : integer


##There are no mixed-type columns

In [21]:
# Check for missing values
df_customers.isnull().sum()

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Number_of_Dependants        0
Familial_Status             0
income                      0
dtype: int64

##There are 11,259 missing values in the First Name column.

In [22]:
#Create a new dataframe to view missing values
df_nan = df_customers[df_customers['First Name'].isnull()==True]

In [23]:
df_nan

user_id First Name Last Name  Gender         State  Age  \
53       76659        NaN   Gilbert    Male      Colorado   26   
73       13738        NaN     Frost  Female     Louisiana   39   
82       89996        NaN    Dawson  Female        Oregon   52   
99       96166        NaN   Oconnor    Male      Oklahoma   51   
105      29778        NaN    Dawson  Female          Utah   63   
...        ...        ...       ...     ...           ...  ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   
206044  200799        NaN  Copeland  Female        Hawaii   52   
206090  167394        NaN     Frost  Female        Hawaii   61   
206162  187532        NaN     Floyd  Female    California   39   
206171  116898        NaN   Delgado  Female      Colorado   23   

        Number_of_Dependants Familial_Status  income  
53                         2         married   41709  
73                         0          single   82518  
82                         3         married  117099  
99                         1         married  155673  
105                        3         married  151819  
...                      ...             ...     ...  
206038                     3         married   87783  
206044                     2         married  108488  
206090                     1         married   45275  
206162                     0          single   56325  
206171                     2         married   59222  

[11259 rows x 9 columns]

In [24]:
df_customers['First Name'].value_counts (dropna = False)

First Name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [25]:
#Addressing missing values
df_customers['First Name'].fillna('Unknown', inplace = True)

In [26]:
#rechecking missing values
df_customers.isnull().sum()

user_id                 0
First Name              0
Last Name               0
Gender                  0
State                   0
Age                     0
Number_of_Dependants    0
Familial_Status         0
income                  0
dtype: int64

In [27]:
#Look for duplicate records
df_dups = df_customers[df_customers.duplicated()]

In [28]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, Number_of_Dependants, Familial_Status, income]
Index: []

##There are no duplicate records in the dataframe. No action required. 

In [29]:
df_customers.shape

(206209, 9)

##Combine df_customers with orders_products_merged_updated_2.pkl 

In [30]:
#Import orders_products_merged_updated_2.pkl
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

In [31]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   order_id                       int64   
 1   user_id                        int64   
 2   order_number                   int64   
 3   orders_day_of_week             int64   
 4   hour_of_day_ordered            int64   
 5   days_since_prior_order         float64 
 6   product_id                     int64   
 7   add_to_cart_order              int64   
 8   reordered                      int64   
 9   product_name                   object  
 10  aisle_id                       int64   
 11  department_id                  int64   
 12  prices                         float64 
 13  _merge                         category
 14  price_range_loc                object  
 15  busiest day                    object  
 16  busiest_period_of_day          object  
 17  max_order                

In [32]:
#Change the data type of user_id to string
df_ords_prods_merge['user_id'] = df_ords_prods_merge['user_id'].astype('str')

In [33]:
#Change the data type of order_id to string
df_ords_prods_merge['order_id'] = df_ords_prods_merge['order_id'].astype('str')

In [34]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   order_id                       object  
 1   user_id                        object  
 2   order_number                   int64   
 3   orders_day_of_week             int64   
 4   hour_of_day_ordered            int64   
 5   days_since_prior_order         float64 
 6   product_id                     int64   
 7   add_to_cart_order              int64   
 8   reordered                      int64   
 9   product_name                   object  
 10  aisle_id                       int64   
 11  department_id                  int64   
 12  prices                         float64 
 13  _merge                         category
 14  price_range_loc                object  
 15  busiest day                    object  
 16  busiest_period_of_day          object  
 17  max_order                

In [35]:
#Drop the _merge column from the df_ords_prods_merge dataframe as it is not needed. 
df_ords_prods_merge = df_ords_prods_merge.drop(columns = ['_merge'])

In [36]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 22 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   order_id                       object 
 1   user_id                        object 
 2   order_number                   int64  
 3   orders_day_of_week             int64  
 4   hour_of_day_ordered            int64  
 5   days_since_prior_order         float64
 6   product_id                     int64  
 7   add_to_cart_order              int64  
 8   reordered                      int64  
 9   product_name                   object 
 10  aisle_id                       int64  
 11  department_id                  int64  
 12  prices                         float64
 13  price_range_loc                object 
 14  busiest day                    object 
 15  busiest_period_of_day          object 
 16  max_order                      int64  
 17  loyalty_flag                   object 
 18  

In [37]:
#Merge df_ords_prods_merge with df_customers
df_ords_prods_all = df_ords_prods_merge.merge(df_customers, on=['user_id'], indicator=True)

In [38]:
#Check the frequency of the new dataframe
df_ords_prods_all['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [39]:
df_ords_prods_all.head(20)

order_id user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329       1             1                   2                    8   
1   2398795       1             2                   3                    7   
2    473747       1             3                   3                   12   
3   2254736       1             4                   4                    7   
4    431534       1             5                   4                   15   
5   3367565       1             6                   2                    7   
6    550135       1             7                   1                    9   
7   3108588       1             8                   1                   14   
8   2295261       1             9                   1                   16   
9   2550362       1            10                   4                    8   
10  2539329       1             1                   2                    8   
11  2539329       1             1                   2                    8   
12  2398795       1             2                   3                    7   
13   473747       1             3                   3                   12   
14  2254736       1             4                   4                    7   
15   431534       1             5                   4                   15   
16  3367565       1             6                   2                    7   
17   550135       1             7                   1                    9   
18  3108588       1             8                   1                   14   
19  2295261       1             9                   1                   16   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     NaN       14084                  2          0   
11                     NaN       12427                  3          0   
12                    15.0       12427                  3          1   
13                    21.0       12427                  2          1   
14                    29.0       12427                  2          1   
15                    28.0       12427                  2          1   
16                    19.0       12427                  2          1   
17                    20.0       12427                  3          1   
18                    14.0       12427                  1          1   
19                     0.0       12427                  6          1   

                               product_name  ...         frequency_flag  \
0                                      Soda  ...  Non-frequent customer   
1                                      Soda  ...  Non-frequent customer   
2                                      Soda  ...  Non-frequent customer   
3                                      Soda  ...  Non-frequent customer   
4                                      Soda  ...  Non-frequent customer   
5                                      Soda  ...  Non-frequent customer   
6                                      Soda  ...  Non-frequent customer   
7                                      Soda  ...  Non-frequent customer   
8                                      Soda  ...  Non-frequent customer   
9                                      Soda  ...  Non-frequent customer   
10  Organic Unsweetened Va

In [40]:
df_ords_prods_all.shape

(32404859, 31)

In [41]:
crosstab = pd.crosstab(df_ords_prods_all['days_since_prior_order'], df_ords_prods_all['order_number'], dropna = False)

In [42]:
crosstab.to_clipboard()

In [43]:
#8. Export this new dataframe as a pickle file.
df_ords_prods_all.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all.pkl'))